### DISCUSSION TITLE
[QUESTION] How to recover from a catastrophic failure?

### DISCUSSION THREAD

**Discussion by wachtell** [longhorn/longhorn#9708] [Q&A] [answered]
After upgrading Traefik from ver 2 to 3 Longhorn lost the CRDs. After reinstalling the CRDs Longhorn came up and running. However, the persistent volumes are in a state of termination/deletion. Is there anyway to recover the PV and re-attach them even though there is no backup of the PV. Can I look on the storage server and find the volume in some library - or is that also deleted?
Thank you in advance!

**Comment by DamiaSan** [comment]
Volumes in Longhorn consist of replicas stored as sparse files on a regular filesystem on each storage node, you can find them in the directory `/var/lib/longhorn` as you can read [here](https://longhorn.io/docs/1.7.2/best-practices/#configuring-default-disks-before-and-after-installation)

**Reply by wachtell** [reply]
Thank you!

**Comment by derekbit** [comment, marked as answer]
@wachtell 

> After upgrading Traefik from ver 2 to 3 Longhorn lost the CRDs.

Can you provide the steps you did for the issue?
Do you know why the upgrade lead to the deletion of Longhorn CRDs?

> the persistent volumes are in a state of termination/deletion.

Can you check if the correspoding Longhorn volumes still exists by `kubectl -n longhorn-system get volumes`?
If yes, you can check the longhorn-manager log and see if there is any clue.

> Can I look on the storage server and find the volume in some library - or is that also deleted?

As the @DamiaSan said, you can find the replica directories of the Longhorn volumes in your Longhorn nodes' disks.
Then, you can rescue your data by mounting https://longhorn.io/docs/1.7.2/advanced-resources/data-recovery/export-from-replica/ and copy the data out.

**Reply by wachtell** [reply]
@derekbit No I am not sure why the Longhorn CRDs got deleted. 
I did run a "kubectl -n longhorn-system get volumes" and found that the volumes are there but in a state of deleting (and otherwise healthy). Is possible to change the state and reattach the PVs?
I found [export from replica](https://longhorn.io/docs/archives/1.1.0/advanced-resources/data-recovery/export-from-replica/) already. However, running "docker run -v /dev:/host/dev -v /proc:/host/proc -v /var/lib/longhorn/replicas/pvc-06b4a8a8-b51d-42c6-a8cc-d8c8d6bc65bc-d890efb2:/volume --privileged longhornio/longhorn-engine:v1.1.0 launch-simple-longhorn pvc-06b4a8a8-b51d-42c6-a8cc-d8c8d6bc65bc 1073741824" I do get a PV on /dev/longhorn/<volume_name>. 
However, I am unsure how best to mount the PV  essentially just cloning it to the longhorn UI. Can you advise?

I found this [migration guide](https://longhorn.io/docs/1.7.2/nodes-and-volumes/volumes/rwx-volumes/#migration-from-previous-external-provisioner), however, I am unsure how to specify the path to the nodes /dev/longhorn/<volume name> and should I use the existing persistent volume claim as the old PVC or just let the old PVC be the new PVC of the new-vol?
Excuse my poor understanding, but I am new to kubernetes.
Thank you in advance

**Comment by DamiaSan** [comment]
@wachtell 
After having created the block device `/dev/longhorn/<volume-name>,  you can mount the block device to get the access to the data, for example to manually write these data into a newly created volume. Is this correct @derekbit ?

**Reply by derekbit** [reply]
Yes, you are right.
